### PR TITLE
oop-user_defined_classes-medium-q2

### PR DESCRIPTION
**Pull request by KaiKhademi**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”

### CODE REVIEW

File: src/oop/user_defined_classes/medium/q2/Car.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    private int horsePower;
+    private int yearMade;
+
+    public Car(String c, int hp, int ym) {
+        colour = c;
+        horsePower = hp;
+        yearMade = ym;
+    }
+
+    public String printSpecs() {
```
Comment (Hedgemon4) [inline src/oop/user_defined_classes/medium/q2/Car.java:14]: Remove this toString

File: src/oop/user_defined_classes/medium/q2/QuestionText.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
+# oop-user_defined_classes-medium-q2
+
+- Java
```
Comment (Hedgemon4) [inline src/oop/user_defined_classes/medium/q2/QuestionText.md:3]:
```suggestion
## Java
```

File: src/oop/user_defined_classes/medium/q2/QuestionText.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+horsePower, and yearMade. This class also has a constructor for 
+initializing the attributes and another method called printSpecs() 
+that displays the colour and horse power of the vehicle.
+
+
+### Car Class
+
+Your `Car` class should have the following items:
+
+- Attributes:
```
Comment (Hedgemon4) [inline src/oop/user_defined_classes/medium/q2/QuestionText.md:16]: Update this so that they know better what they need to code

Comment (Hedgemon4) [reply src/oop/user_defined_classes/medium/q2/QuestionText.md:16]: For example, you could have something more along the lines of:

The `Car` class you are given has the following items: 

whatever you give them

You must add the following items to the class:

x

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/oop/user_defined_classes/medium/q2/QuestionText.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+Little boys love to play with cars and trucks and watch things move. 
+Make a Java class called Car that has attributes called colour, 
+horsePower, and yearMade. This class also has a constructor for 
+initializing the attributes and another method called printSpecs() 
+that displays the colour and horsepower of the vehicle.
+
+
+### Car Class
+
+The `Car` class you are given has the following items:
```
Comment (Hedgemon4) [inline src/oop/user_defined_classes/medium/q2/QuestionText.md:14]: For the medium, you can probably just give them the constructor, and then they can implement a basic method instead

Comment (Hedgemon4) [reply src/oop/user_defined_classes/medium/q2/QuestionText.md:14]: Specifically for this question. For another question, you can give them the attributes, and have them implement the constructor

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/oop/user_defined_classes/medium/q2/QuestionText.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
+# oop-user_defined_classes-medium-q2
+
+## Java
+
+Little boys love to play with cars and trucks and watch things move.
```
Comment (Hedgemon4) [inline src/oop/user_defined_classes/medium/q2/QuestionText.md:5]: I think for this question, we can have them also add the attributes, not just the method.

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/oop/user_defined_classes/medium/q2/QuestionText.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+      with the input values in the constructor and is visible in other classes.
+
+You need to add the following items:
+
+- Attributes:
+    - A `String` called **colour** that is not accessible from other classes
+    - An `int` called **horsePower** that is not accessible from other classes
+    - An `int` called **yearMade** that is not accessible from other classes
+- Methods:
+    - Create a method called **improve** that doubles the **horsePower**.
```
Comment (Hedgemon4) [inline src/oop/user_defined_classes/medium/q2/QuestionText.md:27]: Specify return type

File: src/oop/user_defined_classes/medium/q2/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    @ParameterizedTest
+    @MethodSource("methodInputProvider")
+    public void correctVegetableEatingMethod(String colour, int horsePower, int yearMade) throws Throwable {
+        Object[][] arguments = {
+                {colour, String.class},
+                {horsePower, int.class},
+                {yearMade, int.class}
+        };
+        Object instance = classInstance.createInstance(arguments);
+        classInstance.callMethod("improve", instance);
```
Comment (Hedgemon4) [inline src/oop/user_defined_classes/medium/q2/MainTest.java:86]: Test return type

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/oop/user_defined_classes/medium/q2/MainTest.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        assertTrue(classInstance.hasModifier(varColour, "private"),
+                "Your " + varColour + " attribute does not have the correct visibility modifier.");
+        assertTrue(classInstance.hasField(varPower),
+                "Your " + className + " class is missing the " + varPower + " attribute, or it is spelled incorrectly.");
+        assertTrue(classInstance.hasField(varPower, int.class),
+                "Your " + varPower + " attribute does not have the correct return type.");
+        assertTrue(classInstance.hasModifier(varPower, "private"),
+                "Your " + varPower + " attribute does not have the correct visibility modifier.");
+        assertTrue(classInstance.hasField(varYear),
+                "Your " + className + " class is missing the " + varYear+ " attribute, or it is spelled incorrectly..");
```
Comment (Hedgemon4) [inline src/oop/user_defined_classes/medium/q2/MainTest.java:68]:
```suggestion
                "Your " + className + " class is missing the " + varYear+ " attribute, or it is spelled incorrectly.");
```

File: src/oop/user_defined_classes/medium/q2/MainTest.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+                {colour, String.class},
+                {horsePower, int.class},
+                {yearMade, int.class}
+        };
+        Object instance = classInstance.createInstance(arguments);
+        Object output = classInstance.callMethod(methodName, instance);
+        _assertEquals(horsePower * 2, classInstance.getFieldValue(instance, varPower),
+                "Your " + className + " improve method does not doubles the " + varPower + ".");
+        assertNull(output, "Your " + methodName + " method must not return anything.");
+    }
```
Comment (Hedgemon4) [inline src/oop/user_defined_classes/medium/q2/MainTest.java:96]: You also need to check the visibility modifier.

File: src/oop/user_defined_classes/medium/q2/MainTest.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        Object[][] arguments = {
+                {colour, String.class},
+                {horsePower, int.class},
+                {yearMade, int.class}
+        };
+        Object instance = classInstance.createInstance(arguments);
+        Object output = classInstance.callMethod(methodName, instance);
+        _assertEquals(horsePower * 2, classInstance.getFieldValue(instance, varPower),
+                "Your " + className + " improve method does not doubles the " + varPower + ".");
+        assertNull(output, "Your " + methodName + " method must not return anything.");
```
Comment (Hedgemon4) [inline src/oop/user_defined_classes/medium/q2/MainTest.java:95]: Use the return type assert, not a null assert.

File: src/oop/user_defined_classes/medium/q2/MainTest.java [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+                Arguments.of("Red", 300, 2003),
+                Arguments.of("Blue", 200, 2014),
+                Arguments.of("Black", 100, 1986),
+                Arguments.of("Yellow", 250, 2001)
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("methodInputProvider")
+    public void correctVegetableEatingMethod(String colour, int horsePower, int yearMade) throws Throwable {
```
Comment (Hedgemon4) [inline src/oop/user_defined_classes/medium/q2/MainTest.java:85]: Rename this to correspond to the question

File: src/oop/user_defined_classes/medium/q2/MainTest.java [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+                "Your " + varPower + " attribute does not have the correct visibility modifier.");
+        assertTrue(classInstance.hasField(varYear),
+                "Your " + className + " class is missing the " + varYear+ " attribute, or it is spelled incorrectly..");
+        assertTrue(classInstance.hasField(varYear, int.class),
+                "Your " + varYear + " attribute does not have the correct return type.");
+        assertTrue(classInstance.hasModifier(varYear, "private"),
+                "Your " + varYear + " attribute does not have the correct visibility modifier.");
+    }
+    private static Stream<Arguments> methodInputProvider() {
+        return Stream.of(
```
Comment (Hedgemon4) [inline src/oop/user_defined_classes/medium/q2/MainTest.java:75]: You should add a case for zero and 1

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/oop/user_defined_classes/medium/q2/MainTest.java [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+            _assertEquals(horsePower, classInstance.getFieldValue(instance, varPower), modifiedClassMessage);
+            _assertEquals(yearMade, classInstance.getFieldValue(instance, varYear), modifiedClassMessage);
+            _assertEquals(colour, classInstance.getFieldValue(instance, varColour), modifiedClassMessage);
+        }
+    }
+    @Test
+    public void classHasRequiredFields() {
+        assertTrue(classInstance.hasField(varColour),
+                "Your " + className + " class is missing the " + varColour + " attribute, or it is spelled incorrectly.");
+        assertTrue(classInstance.hasField(varColour, String.class),
```
Comment (Hedgemon4) [inline src/oop/user_defined_classes/medium/q2/MainTest.java:57]: You are not checking that the varColour attribute exists

File: src/oop/user_defined_classes/medium/q2/MainTest.java [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+    public void correctImproveMethod(String colour, int horsePower, int yearMade) throws Throwable {
+        Object[][] arguments = {
+                {colour, String.class},
+                {horsePower, int.class},
+                {yearMade, int.class}
+        };
+        Object instance = classInstance.createInstance(arguments);
+        Object output = classInstance.callMethod(methodName, instance);
+        _assertEquals(horsePower * 2, classInstance.getFieldValue(instance, varPower),
+                "Your " + className + " improve method does not doubles the " + varPower + ".");
```
Comment (Hedgemon4) [inline src/oop/user_defined_classes/medium/q2/MainTest.java:107]:
```suggestion
                "Your " + className + " improve method does not double the value of the " + varPower + "  attribute.");
```

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/oop/user_defined_classes/medium/q2/MainTest.java (hunk 13 above)
Comment (KaiKhademi) [reply src/oop/user_defined_classes/medium/q2/MainTest.java:57]: It is checking if the field exists I don't get the issue

**Review by Hedgemon4 — APPROVED** [review]
lgtm

**Pull request merged**